### PR TITLE
Add the uninstall command

### PR DESCRIPTION
**Pull request by mlorbetske**
Add support for uninstall
Add logic to release mount points (so that uninstalls can still happen after something in the thing being uninstalled has been used)
Add stdout to restore failed output
Skip downloading NuGet.exe if it's already present

### CODE REVIEW

**Comment by sayedihashimi** [comment]
+1

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 1]
```diff
@@ -662,6 +667,90 @@ private CreationResultStatus EnterInstallFlow()
             return CreationResultStatus.Success;
         }
 
+        private void ProcessUninstallRequests(IEnumerable<string> uninstallRequests, bool warnOnFailure)
```
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: I'd move this to a separate class - somewhere outside of New3Command.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 2]
```diff
@@ -679,6 +768,18 @@ private CreationResultStatus EnterMaintenanceFlow()
                 return CreationResultStatus.InvalidParamValues;
             }
 
+            if (InstallHasValue &&
+                ((Install.Count > 0) && (Install[0] != null)))
+            {
+                ProcessUninstallRequests(Install.Select(x => x.Split(new[] { "::" }, StringSplitOptions.None)[0]), false);
```
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Not sure why uninstalls are happening when the user requested installs.

File: src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs [hunk 3]
```diff
@@ -173,25 +188,48 @@ private void ScanMountPointForTemplatesAndLangpacks(IMountPoint mountPoint, stri
                 {
                     AddTemplateToMemoryCache(template);
                 }
+
+                anythingFound |= templateList.Count > 0;
```
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs]: anythingFound doesn't take into account the found langpacks. I think this should be:

anythingFound |= templateList.Count > 0 || localizationInfo.Count > 0

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 2 above)
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: To remove anything being overridden (upgrade should remove the old package)

Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Makes sense.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 1 above)
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Moving to `Installer`

File: src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs (hunk 3 above)
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Edge/Settings/TemplateCache.cs]: Done

**Pull request merged**